src/Microsoft.AspNetCore.SignalR.Client/HubConnection.cs

```diff
@@ -122,60 +120,62 @@ public void On(string methodName, Type[] parameterTypes, Action<object[]> handle
                 }
-                _pendingCalls.Add(descriptor.Id, irq);
+                _pendingCalls.Add(invocationMessage.InvocationId, irq);
             }
 
             // Trace the invocation, but only if that logging level is enabled (because building the args list is a bit slow)
             if (_logger.IsEnabled(LogLevel.Trace))
             {
                 var argsList = string.Join(", ", args.Select(a => a.GetType().FullName));
-                _logger.LogTrace("Invocation #{0}: {1} {2}({3})", descriptor.Id, returnType.FullName, methodName, argsList);
+                _logger.LogTrace("Invocation #{0}: {1} {2}({3})", invocationMessage.InvocationId, returnType.FullName, methodName, argsList);
             }
 
             try
             {
-                var ms = new MemoryStream();
-                await _adapter.WriteMessageAsync(descriptor, ms, cancellationToken);
+                var payload = _protocol.WriteMessage(invocationMessage);
 
-                _logger.LogInformation("Sending Invocation #{0}", descriptor.Id);
+                _logger.LogInformation("Sending Invocation #{0}", invocationMessage.InvocationId);
 
                 // TODO: Format.Text - who, where and when decides about the format of outgoing messages
-                await _connection.SendAsync(ms.ToArray(), MessageType.Text, cancellationToken);
-                _logger.LogInformation("Sending Invocation #{0} complete", descriptor.Id);
+                await _connection.SendAsync(payload, MessageType.Text, cancellationToken);
+                _logger.LogInformation("Sending Invocation #{0} complete", invocationMessage.InvocationId);
             }
             catch (Exception ex)
             {
-                _logger.LogError(0, ex, "Sending Invocation #{0} failed", descriptor.Id);
+                _logger.LogError(0, ex, "Sending Invocation #{0} failed", invocationMessage.InvocationId);
                 irq.Completion.TrySetException(ex);
                 lock (_pendingCallsLock)
                 {
-                    _pendingCalls.Remove(descriptor.Id);
+                    _pendingCalls.Remove(invocationMessage.InvocationId);
                 }
             }
 
             // Return the completion task. It will be completed by ReceiveMessages when the response is received.
             return await irq.Completion.Task;
         }
 
-        private async void OnDataReceived(byte[] data, MessageType messageType)
+        private void OnDataReceived(byte[] data, MessageType messageType)
         {
-            var message
-                = await _adapter.ReadMessageAsync(new MemoryStream(data), _binder, _connectionActive.Token);
+            if (!_protocol.TryParseMessage(data, _binder, out var message))
+            {
+                _logger.LogError("Received invalid message");
+                throw new InvalidOperationException("Received invalid message");
```

Comment:
It would depend on what the caller will do with the exception. I can avoid throwing here, but I also think that IConnection shouldn't allow an exception in the Received event bring down the process.